### PR TITLE
fix(AI): Fix general behavior of reverse only thruster escorts

### PR DESCRIPTION
**Pull request by MagicMuscleMan**
**Bugfix:** This PR addresses issue #{{6604}}.

## Fix Details
This is an extension of the partial solution provided in [pull request #8336](https://github.com/endless-sky/endless-sky/pull/8336).
With it the flagship and escorts should behave reasonably (although not perfect) with a reverse thruster as their only or main thruster.

The acceleration and maximum speed functions are structured (forward with/without afterburner, reverse or use whatever is faster).

Based on that work, the AI is optimized to reasonably handle cases where the reverse thruster is stronger than the forward thruster, especially moving, attacking, circling and keeping station (escorting the player).

## Testing Done

- [x]     Order reverse thruster only escorts to "gather"
- [x]     Order reverse thrusteronly escorts to "hold"
- [x]     Let reverse thruster only escorts jump
- [x]     Let reverse thruster only escorts use hyperspace
- [x]     Let reverse thruster only escorts use a wormhole
- [x]     Let reverse thruster only escorts land on a planet
- [x]     Let reverse thruster only escorts refuel on a planet
- [x]     Let reverse thruster only escorts escort the player inside the invisible fence
- [x]     Let reverse thruster only escorts escort the player outside the invisible fence
- [x]     Let reverse thruster only escorts fight
- [x]     Let reverse thurster only flagship land
- [x]     Let reverse thurster only flagship use a 
- [x]     Let reverse thurster only flagship jump
- [x]     Let reverse thurster only flagship use hyperspace
- [x]     Use forward thruster only escorts
- [x]     Use forward and reverse thruster flagship

### CODE REVIEW

**Comment by MagicMuscleMan** [comment]
Does anyone know, why the code style check fails? I only see an empty error message, but maybe I am holding it wrong.

File: source/Ship.cpp
Comment (warp-core) [inline source/Ship.cpp:3541]:
```suggestion
	if(allThrust)
```

Comment (MagicMuscleMan) [reply source/Ship.cpp]: Thanks @warp-core! Turned out, that my browser filtered out the necessary JavaScript which included the information.

**Comment by warp-core** [comment]
I haven't looked closely at the code here, but it seems like this PR (and your other one) might be doing something similar to 7703, just in a different way.

**Comment by MagicMuscleMan** [comment]
> I haven't looked closely at the code here, but it seems like this PR (and your other one) might be doing something similar to 7703, just in a different way.

Yes, there are some overlaps, but in addition they both seem to solve things the other one does not. I think it's possible to merge either one and then adapt the other one to combine the best of both worlds.
If you prefer, I think @wjp can also create a unified version before, but this might increase two already not so small PRs to an even larger one.

**Comment by MagicMuscleMan** [comment]
Just to be sure: I can't fix the unsuccessful check for the Linux Appimage creation, as this is an infrastructure problem. Right?

File: source/Ship.h
Comment (RisingLeaf) [inline source/Ship.h:358]:
```suggestion
	double ForwardThrust(bool allThrust = false) const;
	double ForwardAcceleration(bool allThrust = false) const;
```

File: source/Ship.h
Comment (RisingLeaf) [inline source/Ship.h:364]:
```suggestion
	double Acceleration(bool allThrust = false) const;
	bool PrefersForwardAcceleration(bool allThrust = false) const;
	double MaxForwardSpeed(bool allThrust = false) const;
	double MaxReverseSpeed() const;
	double MaxSpeed(bool allThrust = false) const;
```

false seems to be the most common case, like this it will save some changes

Comment (RisingLeaf) [reply source/Ship.h:364]: obviously also remove false everywhere

File: source/Ship.cpp
Comment (RisingLeaf) [inline source/Ship.cpp:3580]:
```suggestion
	// v * drag / mass = thrust / mass
	// v * drag = thrust
	// v = thrust / drag
```

why boolean operations in comment XD

File: source/Ship.cpp
Comment (RisingLeaf) [inline source/Ship.cpp:3540]:
```suggestion
	double thrust = allThrust ? attributes.Get("thrust") :  attributes.Get("thrust") + attributes.Get("afterburner thrust");
	return thrust ? thrust : attributes.Get("afterburner thrust");
```

I would prefer it like this

File: source/Ship.cpp
Comment (RisingLeaf) [inline source/Ship.cpp:3544]:
```suggestion
```

File: source/Fleet.cpp
Comment (RisingLeaf) [inline source/Fleet.cpp:529]: good catch, velocity is a vector, so I like this name change
but if we change it here, it should also be changed above

File: source/Fleet.cpp
Comment (RisingLeaf) [inline source/Fleet.cpp:473]: change this to speed if you change the other one to speed

File: source/AI.cpp
Comment (RisingLeaf) [inline source/AI.cpp:2915]: same here

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

File: source/Ship.h
Comment (MagicMuscleMan) [reply source/Ship.h:358]: I initially wanted to implement it as you suggested as using an afterburner clearly should be the exception when having a regular thruster. Then I hesitated and wanted to check the users of that function first.
While doing so I learned two things:

- There is value in being explicitly forced to think about whether in a specific situation an additional afterburner should be used
- Although most of the (in-game) time an additional afterburner won't be used, the majority of AI logic might still use an afterburner as the special cases typically include some urgency

Do you still think, that I should commit the change proposed by you?

File: source/Ship.h
Comment (MagicMuscleMan) [reply source/Ship.h:364]: @RisingLeaf see above

File: source/Ship.cpp
Comment (MagicMuscleMan) [reply source/Ship.cpp:3540]: I see where you want to go with that. I think it's a bit harder to grasp, as you now have 4 code paths (although one trivial one) instead of the needed 3 code paths. But I am fine with it if you prefer that.
But I guess the double space after the colon in your suggestion is not intended or do I miss something?

File: source/Ship.cpp
Comment (MagicMuscleMan) [reply source/Ship.cpp:3544]: Waiting on answer on comment directly above.

File: source/Ship.cpp
Comment (RisingLeaf) [reply source/Ship.cpp:3540]: lets wait on another reviewer with this one, I am not too sure now.

File: source/Ship.h
Comment (RisingLeaf) [reply source/Ship.h:358]: same here, I think it would be the right way, but I dont want you to do something only to revert it later so lets wait on another opinion

File: source/Fleet.cpp
Comment (MagicMuscleMan) [reply source/Fleet.cpp:529]: I can return the compliment: good catch!

File: source/Fleet.cpp
Comment (MagicMuscleMan) [reply source/Fleet.cpp]: done

File: source/AI.cpp
Comment (MagicMuscleMan) [reply source/AI.cpp]: Done, thanks

**Comment by MagicMuscleMan** [comment]
Thanks for the review! Should I do a rebase to more logically structure the commit history after all conversations are resolved?

**Comment by Amazinite** [comment]
We squash merge all PRs, so no need to rebase.
Heads up that we're currently comparing this PR to #7703 to determine which one we want, as we can't have both, at least not easily.

**Comment by Amazinite** [comment]
Marking both #7703 and #8354 as orphaned and closing them. There's a lot of agreement that the reverse thruster AI needs improved upon, and both of these PRs do make improvements upon it, but I think they each miss the mark in different ways. Therefore, a combination of the two is what we should be looking to implement, but that should be done in a single PR rather than being split across two PRs, especially since some of the changes aren't compatible with one another. Because of that, I think it's better to close these out rather than letting them sit.

I've planned to combine these PRs myself, but I haven't gotten around to sinking my teeth into them just yet. If anyone else starts working on them, please let me know and I'll send over anything I have if I've started but don't think I'll be finishing soon.

**Pull request closed without merging**
